PopulateTools/gobierto
# Gobierto People: Person management front-end

Connects to #190.

### Review notes

These changes depend on https://github.com/PopulateTools/gobierto/pull/194. Let's try to get it merged first to keep a significantly better context here.

### What does this PR do?

This PR takes the back-end introduced in #194 and integrates most of the front-end stuff from sandbox files in it. This is including markup, styles and interactions.

First, the Content Block Record UI has been completely redefined according to the spec. All operations such as Add new record, Edit record, Delete record or Cancel changes while editing are now being handled client-side. All those interactions are being tested properly since the most relevant integration test cases have been updated to execute JS and to cover more features. It also reimplements the Content Block management UI itself (Add or delete Block, manage columns in Block...). It now exposes new styles and normalised interactions across sections.

Secondly, It extends the Person form by adding a couple new attributes that were required in the latest spec:

- `avatar_url`, and file upload management.
- `bio_url`, and file upload management.

Also, it adds a policy class to determine which actions concerning Content Blocks are allowed per Admin user.

Last but not least, we have integrated the [Basecamp's Trix](https://github.com/basecamp/trix) WYSIWYG editor into the app, which includes:

- A `FormBuilder` helper that can be used to abstract the markup behind the rich text editor user control. It allows specifying a custom path for handling file attachments:

```ruby
f.rich_text_area :wadus, attachment_path: admin_wadus_file_attachments_path
```

- A custom implementation for handling file uploads through the user control. We're storing these files as attachments in our main asset store (S3 bucket) for now.

### How should this be manually tested?

Having http://gobierto.dev/admin/people/people as the main entry point for this feature, let's check that:

- [x] The Person management form looks and behaves like the spec/sandbox.
- [x] An Avatar image can be uploaded at Person level, and it's properly shown below the input field when available.
- [x] A generic attachment (PDF, although we're supporting any type of file) can be uploaded at Person level, and its URL to download is properly shows below the input field when available.

![screen shot 2017-01-02 at 4 38 58 pm](https://cloud.githubusercontent.com/assets/126392/21592077/9b81e908-d10a-11e6-8e40-71605cbe2379.jpg)

Regarding Content Block management, here's the check-list:

- [x] A new Record can be added to any of the Content Blocks.
- [x] A Record from any Content Block can be Edited.
- [x] Changes are discarded when hitting the "Cancel" link, in any case.
- [x] When editing Content Blocks, columns can be added or deleted in all available languages (ES, CA at the time being).
- [x] Regular Admins aren't allowed to Create, Edit or Delete Content Blocks.
- [x] Managing Admins (Manager Admin and God Admin) are allowed to do whatever they want to Content Blocks.

These may look like:

![screen shot 2017-01-02 at 4 47 52 pm](https://cloud.githubusercontent.com/assets/126392/21592145/3c09dc6e-d10b-11e6-8102-fa7ea36fc382.jpg)

![screen shot 2017-01-02 at 5 11 42 pm](https://cloud.githubusercontent.com/assets/126392/21592540/96866218-d10e-11e6-8ab4-bc5a0d9f1ce0.jpg)

Finally, regarding the WYSIWYG user control:

- [x] Person's bio can be edited in rich text format.
- [x] Any kind of images can be dropped or pasted there and they are automatically uploaded to our main store.

![screen shot 2017-01-03 at 8 00 53 am](https://cloud.githubusercontent.com/assets/126392/21601506/b8cbd952-d18b-11e6-8d94-cee8d031d8fe.jpg)
